### ISSUE TITLE
Release 1.4.0 drawElements WebGL warning (causes flickering)

### ISSUE THREAD

**Issue by MeisterGit** [playcanvas/engine#1229]
Error: WebGL warning: drawElements: Vertex fetch requires vertex #3, but attribs only supply 3. playcanvas-stable.dbg.js:5608:9

With zero code changes, if I revert to PC 1.3.0 the issue vanishes.

The warning is showing up when I update text character-by-character in a text box. The text element itself is in a custom layer.  This blackout only occurs when my text is actively advancing itself, and is otherwise fine.

The visible manifestation of this issue is a single plane in my world layer only drawing 50% of its vertices on frames where my text element's text attribute updated.

The top right triangle renders, bottom left triangle of the plane is gone.

I figured this strange knock-on effect was worth an issue, since I'm seeing a glitch with the rendering of Entity A in the World Layer while only adjusting Entity B in a Custom Layer.

**Comment by MeisterGit** [comment]
Upon further investigation, I have tracked down the exact changes which trigger this behavior!

 "No longer needlessly nuke texture unit and bound buffer cache."
commit 06a96fa7cd9566a2e7f92e866b71f8abaa298dc3

Is the cause. If I manually restore:

`					for (var i = 0;i < 16;i++) {		
					  this.textureUnits[i] = null;		
					}`

and

`				this.boundBuffer = null;		
				this.indexBuffer = null;`

Then the flickering vanishes, and the WebGL warning goes away.

**Comment by MeisterGit** [comment]
Upon further investigation, all that needs to be restored is:

`this.indexBuffer = null;`

You can continue to cull out the bound buffer null and the texture units null in my case.

**Comment by willeastcott** [comment]
That is very odd. What platform/browser are you observing this with? Does it happen in other browsers too?

**Comment by MeisterGit** [comment]
Windows 10, the issue occurs in both Firefox and Chrome

**Comment by willeastcott** [comment]
Ooo. I wonder whether the problem is because of this:

https://github.com/playcanvas/engine/blob/master/src/graphics/index-buffer.js#L146

If you monkey patch that function and add the line:

```
this.device.indexBuffer = null;
```

...does that fix the problem?

**Comment by MeisterGit** [comment]
Close! The entire plane now flickers black instead of half, but it only happens on the first character of each text animation update.

Not perfect like restoring the null to its original position, but you're going in the right direction.

**Comment by MeisterGit** [comment]
Correction:

I have to restore both this.indexBuffer AND this.boundBuffer to null in device.js to get a flawless 0 flicker result. When I returned to double check this.indexBuffer alone, it had the same single-flicker going on as the current monkey patch solution.

**Comment by willeastcott** [comment]
So what if you add this to a script if your project:

```
var oldIndexBufferUnlock = pc.IndexBuffer.prototype.unlock;
var oldVertexBufferUnlock = pc.VertexBuffer.prototype.unlock;

pc.IndexBuffer.prototype.unlock = function () {
    oldIndexBufferUnlock();
    this.device.indexBuffer = null;
};
pc.VertexBuffer.prototype.unlock = function () {
    oldVertexBufferUnlock();
    this.device.vertexBuffer = null;
};
```

**Comment by MeisterGit** [comment]
Tried it! Same result. A single flicker occurs with the start of each new message displayed.

As an aside, this solution has two WebGL messages instead of one:

Error: WebGL warning: drawElements: Vertex fetch requires vertex #3, but attribs only supply 3.
playcanvas-stable.dbg.js:5608:9
Error: WebGL warning: drawElements: Vertex fetch requires vertex #3, but attribs only supply 2.
playcanvas-stable.dbg.js:5608:9

**Comment by willeastcott** [comment]
Can I access this scene to reproduce what you're seeing?

**Comment by MeisterGit** [comment]
```
    pc.IndexBuffer.prototype.oldIndexBufferUnlock = pc.IndexBuffer.prototype.unlock;
    pc.VertexBuffer.prototype.oldVertexBufferUnlock = pc.VertexBuffer.prototype.unlock;

    pc.IndexBuffer.prototype.unlock = function () {
        this.oldIndexBufferUnlock();
        this.device.indexBuffer = null;
        this.device.boundBuffer = null;
    };
    pc.VertexBuffer.prototype.unlock = function () {
        this.oldVertexBufferUnlock();
        this.device.vertexBuffer = null;
        this.device.boundBuffer = null;
    };
```

This modification appears to be working. Testing some more.

**Comment by willeastcott** [comment]
Interesting. 👍

**Comment by MeisterGit** [comment]
I reproduced the issue in what I believe is a minimal example! Added you to the team on it. "Textbox Bug"

The name isn't quite representative of what's going on. I learned more about the issue while reproducing it.

My camera has a 'fader' in front of it, a plane which I tweak the opacity on to fade in and out of scenes. If this fader plane is disabled, the issue goes away, but when enabled, even if 100% invisible, the bug triggers the plane to partially render.

The monkey patch is in the code for messagebox.js, commented out. If you comment it back in, the issue vanishes as expected.
